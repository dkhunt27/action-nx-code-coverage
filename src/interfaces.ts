import {
  FileListType,
  FinalFileListType,
  GistUpdateFileType,
  JcsMergedType,
  JcsParsedType,
  SummaryFileListType
} from './types'

export interface MainInputs {
  coverageRan: boolean
  coverageFolder: string
  coverageBaseFolder: string
  token: string
  githubWorkspace: string
  gistId: string
  gistToken: string
}

export interface UpsertCommentInputs {
  token: string
  prNumber: number
  body: string
  hiddenHeader: string
  repoOwner: string
  repoRepo: string
}

export interface ProcessCoverageFilesInputs {
  workspacePath: string
  coverageFolder: string
  coverageBaseFolder: string
}

export interface BuildFileListInputs {
  workspacePath: string
  folder: string
}

export interface ListCoverageFilesInputs {
  fileToFind: string
  parseFileFn: (_: string) => Promise<JcsParsedType | string>
  workspacePath: string
  initDir: string
  dir?: string
  results?: FileListType[]
}

export interface BuildCommentInputs {
  results: JcsMergedType[]
  compact?: boolean
}

export interface MergeFileListsInputs {
  summaryFileList: SummaryFileListType[]
  baseSummaryFileList: SummaryFileListType[]
  finalFileList: FinalFileListType[]
}

export interface CreateCoverageBadgeInputs {
  label: string
  message: string
}

export interface UpdateCoverageGistInputs {
  files: GistUpdateFileType
  gistId: string
  gistToken: string
}
