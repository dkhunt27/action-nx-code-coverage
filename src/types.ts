export type TabulateOptionsType = {
  repository: string
  commit: string
  prefix: string
  head: string
  base: string
}

export type ParsedContextType = {
  repositoryFullName: string
  pullRequestHeadSha: string
  pullRequestHeadRef: string
  pullRequestBaseRef: string
  pullRequestNumber: number
  repoOwner: string
  repoRepo: string
}

export type JcsFileListType = {
  parsed: JcsParsedType
  path: string
  fullPath: string
  relativePath: string
  app: string
}

export type JcsDataType = {
  total: number
  covered: number
  skipped: number
  pct: number | 'Unknown'
}

export type JcsSummaryType = {
  lines: JcsDataType
  functions: JcsDataType
  statements: JcsDataType
  branches: JcsDataType
}

export type JcsParsedType = Record<string, JcsSummaryType>

export type JcsMergedItemType = {
  app: string
  parsedTotal: JcsSummaryType
  parsedOthers: Record<string, JcsSummaryType>
}

export type JcsMergedType = {
  app: string
  coverage: number
  base: number | null
  diff: number | null
  details: string
}

export type FinalFileListType = {
  app: string
  parsed: string
}

export type SummaryFileListType = {
  app: string
  parsed: JcsParsedType
}

export type FileListType = {
  app: string
  parsed: JcsParsedType | string
}

export type GistContentType = {
  schemaVersion: number
  label: string
  message: string
  labelColor?: string
  color?: string
  isError?: string
  namedLogo?: string
  logoSvg?: string
  logoColor?: string
  logoWidth?: number
  logoPosition?: string
  style?: string
  cacheSeconds?: number
}

export type GistUpdateFileType = {
  [key: string]: Partial<{[key: string]: unknown}>
}
