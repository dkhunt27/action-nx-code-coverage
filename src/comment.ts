import {details, fragment, summary, table, tbody, th, tr} from './html'
import {BuildCommentInputs} from './interfaces'
import {tabulate} from './tabulate'

const renderEmoji = (diff: number): string => {
  if (diff < 0) return '❌'

  return '✅'
}

export const buildComment = ({
  results,
  hideCoverageReports,
  hideUnchanged
}: BuildCommentInputs): string => {
  const rows = results
    .map(result => {
      let plus = ''
      let arrow = ''
      let diffHtml = ''

      // when no tests, not sure if output is undefined or 'Unknown'; TODO: add test case
      if (
        result.diff !== undefined &&
        result.diff !== null &&
        (result.diff as unknown) !== 'Unknown'
      ) {
        if (result.diff < 0) {
          arrow = '▾'
        } else if (result.diff > 0) {
          plus = '+'
          arrow = '▴'
        }

        diffHtml = th(
          renderEmoji(result.diff),
          ' ',
          arrow,
          ' ',
          plus,
          result.diff.toFixed(2),
          '%'
        )
      }

      // when no tests, not sure if is undefined or 'Unknown'; TODO: add test case
      let coverage
      if (
        result.coverage === undefined ||
        result.coverage === null ||
        (result.coverage as unknown) === 'Unknown'
      ) {
        coverage = 'unknown'
      } else {
        coverage = result.coverage.toFixed(2)
      }

      if (result.diff === 0 && hideUnchanged) {
        return {
          row: '',
          htmlResults: ''
        }
      } else {
        const htmlResults = tabulate(result.details)
        return {
          row: tr(
            th(result.app.replace(/^\/+/g, '')),
            th(coverage, '%'),
            diffHtml
          ),
          htmlResults
        }
      }
    })
    .filter(row => row.row !== '')

  const tableHtml = hideCoverageReports
    ? table(tbody(rows.map(row => row.row).join('')))
    : rows
        .map(
          row =>
            `${table(tbody(row.row))} \n\n ${details(summary('Coverage Report'), row.htmlResults)} <br/>`
        )
        .join('')

  const title = `<h2>Code Coverage</h2>`
  return fragment(title, tableHtml)
}
