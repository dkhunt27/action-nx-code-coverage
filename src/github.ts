import {context, getOctokit} from '@actions/github'
import {debug as logDebug, info as logInfo} from '@actions/core'
import {ParsedContextType} from './types'
import {UpsertCommentInputs} from './interfaces'

// Every comment written by our action will have this hidden
// header on top, and will be used to identify which comments
// to update/delete etc
const appendHiddenHeaderToComment = (
  body: string,
  hiddenHeader: string
): string => hiddenHeader + body

export const buildParsedContext = (): ParsedContextType => {
  logDebug(`context: ${JSON.stringify(context, null, 2)}`)

  if (!context.payload.repository) {
    throw new Error('context.payload.repository cannot be null')
  }

  const parsedContext: ParsedContextType = {
    repositoryFullName: context.payload.repository.full_name as string,
    pullRequestHeadSha: context.payload.pull_request
      ? context.payload.pull_request.head.sha
      : '',
    pullRequestHeadRef: context.payload.pull_request
      ? context.payload.pull_request.head.ref
      : '',
    pullRequestBaseRef: context.payload.pull_request
      ? context.payload.pull_request.base.ref
      : '',
    pullRequestNumber: context.payload.pull_request
      ? context.payload.pull_request.number
      : -1,
    repoOwner: context.repo.owner,
    repoRepo: context.repo.repo
  }

  logInfo(`parsedContext: ${JSON.stringify(parsedContext, null, 2)}`)
  return parsedContext
}

export const upsertComment = async ({
  token,
  prNumber,
  body,
  hiddenHeader,
  repoOwner,
  repoRepo
}: UpsertCommentInputs): Promise<void> => {
  const github = getOctokit(token).rest

  // find previous comment made by this action
  const {data: existingComments} = await github.issues.listComments({
    issue_number: prNumber,
    owner: repoOwner,
    repo: repoRepo
  })

  // find previous comments made by this action
  const coverageComments = existingComments.filter(item =>
    item.body?.startsWith(hiddenHeader)
  )

  // remove the last one from the list, this is the one we will update
  const lastComment = coverageComments.pop()

  // delete all existingComments (only need to keep one, the last one)
  const deletePromises = coverageComments.map(async item =>
    github.issues.deleteComment({
      owner: repoOwner,
      repo: repoRepo,
      comment_id: item.id
    })
  )

  await Promise.all(deletePromises)

  if (lastComment) {
    github.issues.updateComment({
      owner: repoOwner,
      repo: repoRepo,
      comment_id: lastComment.id,
      body: appendHiddenHeaderToComment(body, hiddenHeader)
    })
  } else {
    github.issues.createComment({
      owner: repoOwner,
      repo: repoRepo,
      issue_number: prNumber,
      body: appendHiddenHeaderToComment(body, hiddenHeader)
    })
  }
  return
}
