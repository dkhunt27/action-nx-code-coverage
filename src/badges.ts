import {CreateCoverageBadgeInputs, UpdateCoverageGistInputs} from './interfaces'
import {GistContentType, GistUpdateFileType, JcsMergedType} from './types'
import {getInput} from '@actions/core'
import {getOctokit} from '@actions/github'

// copied from https://github.com/Schneegans/dynamic-badges-action

export const buildGistCoverageFileList = (
  results: JcsMergedType[]
): GistUpdateFileType => {
  let files: GistUpdateFileType = {}
  for (const result of results) {
    const file = createCoverageBadge({
      label: result.app,
      message: result.coverage.toFixed(2)
    })
    files = {...files, ...file}
  }

  return files
}

export const createCoverageBadge = ({
  label,
  message
}: CreateCoverageBadgeInputs): GistUpdateFileType => {
  try {
    // This object will be stringified and uploaded to the gist. The
    // schemaVersion, label and message attributes are always required. All others
    // are optional and added to the content object only if they are given to the
    // action.
    const content: GistContentType = {
      schemaVersion: 1,
      label,
      message
    }

    // Get all optional attributes and add them to the content object if given.
    const labelColor = getInput('label-color')
    const color = getInput('color')
    const isError = getInput('is-error')
    const namedLogo = getInput('named-logo')
    const logoSvg = getInput('logo-svg')
    const logoColor = getInput('label-color')
    const logoWidth = getInput('logo-width')
    const logoPosition = getInput('logo-position')
    const style = getInput('style')
    const cacheSeconds = getInput('cache-seconds')

    if (labelColor !== '') {
      content.labelColor = labelColor
    }

    if (color !== '') {
      content.color = color
    }

    if (isError !== '') {
      content.isError = isError
    }

    if (namedLogo !== '') {
      content.namedLogo = namedLogo
    }

    if (logoSvg !== '') {
      content.logoSvg = logoSvg
    }

    if (logoColor !== '') {
      content.logoColor = logoColor
    }

    if (logoWidth !== '') {
      content.logoWidth = parseInt(logoWidth)
    }

    if (logoPosition !== '') {
      content.logoPosition = logoPosition
    }

    if (style !== '') {
      content.style = style
    }

    if (cacheSeconds !== '') {
      content.cacheSeconds = parseInt(cacheSeconds)
    }

    // For the POST request, the above content is set as file contents for the
    // given filename.

    const fileName = `${label.replace(/\//g, '-')}.json`

    // const request = JSON.stringify({
    //   files: {[`${fileName}`]: {content: JSON.stringify(content)}}
    // })

    return {[fileName]: {content: JSON.stringify(content)}}
  } catch (error) {
    throw error
  }
}

export const updateCoverageGist = async ({
  gistToken,
  gistId,
  files
}: UpdateCoverageGistInputs): Promise<void> => {
  try {
    const github = getOctokit(gistToken).rest

    const res = await github.gists.update({
      gist_id: gistId,
      files
    })

    if (res.status && (res.status < 200 || res.status >= 400)) {
      throw new Error(
        `Failed to create gist, response status code: ${res.status}, status message: ${res.data}`
      )
    }
  } catch (error) {
    throw error
  }
}
