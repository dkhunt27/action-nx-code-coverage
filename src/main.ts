import {buildGistCoverageFileList, updateCoverageGist} from './badges'
import {buildParsedContext, upsertComment} from './github'
import {
  debug as logDebug,
  info as logInfo,
  warning as logWarn
} from '@actions/core'
import {JcsMergedType} from './types'
import {MainInputs} from './interfaces'
import {omit as _omit} from 'lodash'
import {buildComment} from './comment'
import {existsSync} from 'fs'
import {processCoverageFiles} from './json-coverage'

const MAX_GH_COMMENT_SIZE = 65536

export const main = async ({
  coverageRan,
  coverageFolder,
  coverageBaseFolder,
  token,
  githubWorkspace,
  failOnCoverageDecrease,
  gistToken,
  gistId
}: MainInputs): Promise<JcsMergedType[]> => {
  try {
    // hiddenHeader to help identify any previous PR comments
    const hiddenHeaderForCoverage = '<!-- nx-code-coverage -->'
    const hiddenHeaderNoCoverage = '<!-- nx-code-coverage-none -->'
    let commentBody = ''
    let hiddenHeader = ''
    let results: JcsMergedType[] = []

    // check for coverage dir
    const coverageDirExists = existsSync(coverageFolder)

    if (coverageRan && coverageDirExists) {
      logInfo(`Coverage Ran: processing coverage files`)
      results = await processCoverageFiles({
        workspacePath: githubWorkspace,
        coverageFolder,
        coverageBaseFolder
      })

      logInfo(
        `processCoverageFilesResults: ${JSON.stringify(
          _omit(results, 'details'),
          null,
          2
        )}`
      )

      commentBody = buildComment({results})
      if (commentBody.length >= MAX_GH_COMMENT_SIZE) {
        commentBody = buildComment({results, compact: true})
      }
      hiddenHeader = hiddenHeaderForCoverage
    } else {
      logWarn(
        `Coverage Not Ran: NOT processing coverage files ${JSON.stringify({
          coverageRan,
          coverageDirExists
        })}`
      )
      commentBody = 'No coverage ran'
      hiddenHeader = hiddenHeaderNoCoverage
    }

    logDebug(`commentBody: ${commentBody}`)

    const parsedContext = buildParsedContext()

    if (parsedContext.pullRequestNumber !== -1) {
      logInfo(`PR Detected: Updating the PR Comment with Code Coverage`)
      await upsertComment({
        token,
        body: commentBody,
        hiddenHeader,
        prNumber: parsedContext.pullRequestNumber,
        repoOwner: parsedContext.repoOwner,
        repoRepo: parsedContext.repoRepo
      })
    } else {
      // if not a PR, then should be push to main
      // therefore, should always have coverage

      logInfo(`No PR Detected: Updating the Coverage Gist with Code Coverage`)

      const files = buildGistCoverageFileList(results)

      updateCoverageGist({files, gistToken, gistId})
    }

    if (failOnCoverageDecrease && coverageRan && coverageDirExists) {
      const badlyCovered = results.filter(
        result =>
          result.coverage > 0 && result.diff !== null && result.diff < -0.5
      )
      if (badlyCovered.length > 0) {
        throw new Error(
          `Code coverage is decreasing for projects: ${badlyCovered
            .map(p => p.app)
            .join(',')}`
        )
      }
    }

    return results
  } catch (error) {
    throw error
  }
}
