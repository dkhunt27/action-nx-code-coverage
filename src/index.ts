import {
  getBooleanInput,
  getInput,
  error as logError,
  info as logInfo,
  setFailed
} from '@actions/core'
import {MainInputs} from './interfaces'
import {main} from './main'

async function run(): Promise<void> {
  try {
    const noCoverageRan = getBooleanInput('no-coverage-ran')
    const token = getInput('github-token')
    const coverageFolder = getInput('coverage-folder') || 'coverage'
    const coverageBaseFolder =
      getInput('coverage-base-folder') || 'coverage-base'

    const githubWorkspace = process.env.GITHUB_WORKSPACE
    logInfo(`githubWorkspace:  ${githubWorkspace}`)

    if (!githubWorkspace) {
      throw new Error('process.env.GITHUB_WORKSPACE cannot be empty')
    }

    const gistProcessing = getBooleanInput('gist-processing')
    const gistToken = getInput('gist-token', {required: false}) || undefined
    const gistId = getInput('gist-id', {required: false}) || undefined

    if (gistProcessing) {
      if (!gistToken || !gistId) {
        throw new Error(
          'if gistProcessing not false, then gist-token and gist-id must be supplied'
        )
      }
    }
    const mainInputs: MainInputs = {
      coverageRan: !noCoverageRan,
      coverageFolder,
      coverageBaseFolder,
      token,
      githubWorkspace,
      gistProcessing,
      gistToken,
      gistId
    }

    await main(mainInputs)
  } catch (error) {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    setFailed((error as any).message)
  }
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any,github/no-then
run().catch((err: any) => {
  logError(err)
  setFailed(err.message)
})
