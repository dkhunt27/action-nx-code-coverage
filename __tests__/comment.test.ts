/* eslint-disable filenames/match-regex */
import {BuildCommentInputs} from '../src/interfaces'
import {buildComment} from '../src/comment'

describe('comment tests', () => {
  let mockDetails = ''
  beforeEach(() => {
    mockDetails =
      '------------------------------------|---------|----------|---------|---------|-------------------\nFile                                | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s \n------------------------------------|---------|----------|---------|---------|-------------------\nAll files                           | statements |   96.42 | statements |       60 | branches |   96.66 | functions |   96.59 | lines |                   | low \n app                                | statements |   95.83 | statements |      100 | branches |   85.71 | functions |   95.23 | lines |                   | low \n  app.controller.ts                 | statements |   84.61 | statements |      100 | branches |   66.66 | functions |   81.81 | lines | 20-25             | low \n  app.module.ts                     | statements |     100 | statements |      100 | branches |     100 | functions |     100 | lines |                   | medium \n  app.service.ts                    | statements |     100 | statements |      100 | branches |     100 | functions |     100 | lines |                   | medium \n app/interfaces                     | statements |     100 | statements |      100 | branches |     100 | functions |     100 | lines |                   | medium \n  aws-s3-provider.interface.ts      | statements |     100 | statements |      100 | branches |     100 | functions |     100 | lines |                   | medium \n  generate-pdf-service.interface.ts | statements |     100 | statements |      100 | branches |     100 | functions |     100 | lines |                   | medium \n  html-pdf-provider.interface.ts    | statements |     100 | statements |      100 | branches |     100 | functions |     100 | lines |                   | medium \n  lambda-service.interface.ts       | statements |     100 | statements |      100 | branches |     100 | functions |     100 | lines |                   | medium \n  pdf-service.interface.ts          | statements |     100 | statements |      100 | branches |     100 | functions |     100 | lines |                   | medium \n  strapi-service.interface.ts       | statements |     100 | statements |      100 | branches |     100 | functions |     100 | lines |                   | medium \n  template-service.interface.ts     | statements |     100 | statements |      100 | branches |     100 | functions |     100 | lines |                   | medium \n  uuid-provider.interface.ts        | statements |     100 | statements |      100 | branches |     100 | functions |     100 | lines |                   | medium \n app/services/generate-pdf          | statements |     100 | statements |      100 | branches |     100 | functions |     100 | lines |                   | medium \n  generate-pdf.service.ts           | statements |     100 | statements |      100 | branches |     100 | functions |     100 | lines |                   | medium \n app/services/lambda                | statements |   89.47 | statements |       50 | branches |     100 | functions |   91.17 | lines |                   | low \n  lambda.service.ts                 | statements |   89.47 | statements |       50 | branches |     100 | functions |   91.17 | lines | 33,49-51          | low \n app/services/pdf                   | statements |     100 | statements |      100 | branches |     100 | functions |     100 | lines |                   | medium \n  pdf.service.ts                    | statements |     100 | statements |      100 | branches |     100 | functions |     100 | lines |                   | medium \n app/services/strapi                | statements |     100 | statements |      100 | branches |     100 | functions |     100 | lines |                   | medium \n  strapi.service.ts                 | statements |     100 | statements |      100 | branches |     100 | functions |     100 | lines |                   | medium \n app/services/template              | statements |     100 | statements |      100 | branches |     100 | functions |     100 | lines |                   | medium \n  template.service.ts               | statements |     100 | statements |      100 | branches |     100 | functions |     100 | lines |                   | medium \n------------------------------------|---------|----------|---------|---------|-------------------\n'
  })
  describe('buildComment', () => {
    describe('detailed output', () => {
      test('when results empty then empty code coverage comment', () => {
        const input: BuildCommentInputs = {results: []}
        const actual = buildComment(input)
        expect(actual).toStrictEqual('Code Coverage:<p></p>')
      })

      test('when result base/diff null then no diff in code coverage comment', async () => {
        const input: BuildCommentInputs = {
          results: [
            {
              app: 'app A',
              coverage: 50,
              base: null,
              diff: null,
              details: mockDetails
            }
          ]
        }
        const actual = buildComment(input)
        expect(actual).toContain(
          'Code Coverage:<p></p><table><tbody><tr><th>app A</th><th>50.00%</th></tr></tbody></table>'
        )
      })
      test('when result diff is 0 then nothing is displayed in code coverage comment', async () => {
        const input: BuildCommentInputs = {
          results: [
            {
              app: 'app A',
              coverage: 50,
              base: 50,
              diff: 0,
              details: mockDetails
            },
            {
              app: 'app B',
              coverage: 55,
              base: 50,
              diff: 5,
              details: mockDetails
            }
          ]
        }
        const actual = buildComment(input)
        expect(actual).toContain(
          'Code Coverage:<p></p><table><tbody><tr><th>app B</th><th>55.00%</th><th>✅ ▴ +5.00%</th></tr></tbody></table>'
        )
      })
      test('when result diff is negative then negative diff in code coverage comment', async () => {
        const input: BuildCommentInputs = {
          results: [
            {
              app: 'app A',
              coverage: 50,
              base: 75,
              diff: -25,
              details: mockDetails
            }
          ]
        }
        const actual = buildComment(input)
        expect(actual).toContain(
          'Code Coverage:<p></p><table><tbody><tr><th>app A</th><th>50.00%</th><th>❌ ▾ -25.00%</th></tr></tbody></table>'
        )
      })
      test('when result diff is positive then positive diff in code coverage comment', async () => {
        const input: BuildCommentInputs = {
          results: [
            {
              app: 'app A',
              coverage: 50,
              base: 25,
              diff: 25,
              details: mockDetails
            }
          ]
        }
        const actual = buildComment(input)
        expect(actual).toContain(
          'Code Coverage:<p></p><table><tbody><tr><th>app A</th><th>50.00%</th><th>✅ ▴ +25.00%</th></tr></tbody></table>'
        )
      })
      test('renders multiple results', async () => {
        const input: BuildCommentInputs = {
          results: [
            {
              app: 'app A',
              coverage: 50,
              base: 25,
              diff: 25,
              details: mockDetails
            },
            {
              app: 'app B',
              coverage: 50,
              base: 75,
              diff: -25,
              details: mockDetails
            }
          ]
        }
        const actual = buildComment(input)
        expect(actual).toContain(
          'Code Coverage:<p></p><table><tbody><tr><th>app A</th><th>50.00%</th><th>✅ ▴ +25.00%</th></tr></tbody></table>'
        )
        expect(actual).toContain(
          '<table><tbody><tr><th>app B</th><th>50.00%</th><th>❌ ▾ -25.00%</th></tr></tbody></table>'
        )
      })
    })
    describe('compact mode', () => {
      test('when results empty then empty code coverage comment', () => {
        const actual = buildComment({results: [], compact: true})
        expect(actual).toStrictEqual('Code Coverage:<p></p>')
      })

      test('when result base/diff null then no diff in code coverage comment', async () => {
        const input: BuildCommentInputs = {
          results: [
            {
              app: 'app A',
              coverage: 50,
              base: null,
              diff: null,
              details: mockDetails
            }
          ],
          compact: true
        }
        const actual = buildComment(input)
        expect(actual).toContain('<tr><td>app A</td><td>50.00%</td></tr>')
      })
      test('when result diff is 0 then display nothing in code coverage comment', async () => {
        const input: BuildCommentInputs = {
          results: [
            {
              app: 'app A',
              coverage: 50,
              base: 50,
              diff: 0,
              details: mockDetails
            },
            {
              app: 'app B',
              coverage: 55,
              base: 50,
              diff: 5,
              details: mockDetails
            }
          ],
          compact: true
        }
        const actual = buildComment(input)
        expect(actual).toContain(
          '<td>app B</td><td>55.00%</td><td>✅ ▴ +5.00%</td></tr>'
        )
      })
      test('when result diff is negative then negative diff in code coverage comment', async () => {
        const input: BuildCommentInputs = {
          results: [
            {
              app: 'app A',
              coverage: 50,
              base: 75,
              diff: -25,
              details: mockDetails
            }
          ],
          compact: true
        }
        const actual = buildComment(input)
        expect(actual).toContain(
          '<tr><td>app A</td><td>50.00%</td><td>❌ ▾ -25.00%</td></tr>'
        )
      })
      test('when result diff is positive then positive diff in code coverage comment', async () => {
        const input: BuildCommentInputs = {
          results: [
            {
              app: 'app A',
              coverage: 50,
              base: 25,
              diff: 25,
              details: mockDetails
            }
          ],
          compact: true
        }
        const actual = buildComment(input)
        expect(actual).toContain(
          '<tr><td>app A</td><td>50.00%</td><td>✅ ▴ +25.00%</td></tr>'
        )
      })
      test('renders multiple results', async () => {
        const input: BuildCommentInputs = {
          results: [
            {
              app: 'app A',
              coverage: 50,
              base: 25,
              diff: 25,
              details: mockDetails
            },
            {
              app: 'app B',
              coverage: 50,
              base: 75,
              diff: -25,
              details: mockDetails
            }
          ],
          compact: true
        }
        const actual = buildComment(input)
        expect(actual).toContain(
          'Code Coverage:<p></p>' +
            '<table><tbody>' +
            '<tr><th>Project</th><th>Coverage</th><th>Delta</th></tr>' +
            '<tr><td>app A</td><td>50.00%</td><td>✅ ▴ +25.00%</td></tr>' +
            '<tr><td>app B</td><td>50.00%</td><td>❌ ▾ -25.00%</td></tr>' +
            '</tbody></table>'
        )
      })
    })
  })
})
