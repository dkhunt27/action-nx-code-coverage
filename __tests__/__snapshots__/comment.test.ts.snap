// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`comment tests buildComment when hiding coverage reports when hiding unchanged when all results in one then return as expected 1`] = `"<h2>Code Coverage</h2><table><tbody><tr><th>app diff -</th><th>50.00%</th><th>❌ ▾ -25.00%</th></tr><tr><th>app diff +</th><th>50.00%</th><th>✅ ▴ +25.00%</th></tr><tr><th>app diff null</th><th>unknown%</th></tr></tbody></table>"`;

exports[`comment tests buildComment when hiding coverage reports when hiding unchanged when result coverage is undefined then unknown diff in code coverage comment 1`] = `"<h2>Code Coverage</h2><table><tbody><tr><th>app A</th><th>unknown%</th></tr></tbody></table>"`;

exports[`comment tests buildComment when hiding coverage reports when hiding unchanged when result diff is 0 then no code coverage comment 1`] = `"<h2>Code Coverage</h2><table><tbody></tbody></table>"`;

exports[`comment tests buildComment when hiding coverage reports when hiding unchanged when result diff is negative then negative diff in code coverage comment 1`] = `"<h2>Code Coverage</h2><table><tbody><tr><th>app A</th><th>50.00%</th><th>❌ ▾ -25.00%</th></tr></tbody></table>"`;

exports[`comment tests buildComment when hiding coverage reports when hiding unchanged when result diff is positive then positive diff in code coverage comment 1`] = `"<h2>Code Coverage</h2><table><tbody><tr><th>app A</th><th>50.00%</th><th>✅ ▴ +25.00%</th></tr></tbody></table>"`;

exports[`comment tests buildComment when hiding coverage reports when not hiding unchanged when all results in one then return as expected 1`] = `"<h2>Code Coverage</h2><table><tbody><tr><th>app diff 0</th><th>50.00%</th><th>✅  0.00%</th></tr><tr><th>app diff -</th><th>50.00%</th><th>❌ ▾ -25.00%</th></tr><tr><th>app diff +</th><th>50.00%</th><th>✅ ▴ +25.00%</th></tr><tr><th>app diff null</th><th>unknown%</th></tr></tbody></table>"`;

exports[`comment tests buildComment when hiding coverage reports when not hiding unchanged when result coverage is undefined then unknown diff in code coverage comment 1`] = `"<h2>Code Coverage</h2><table><tbody><tr><th>app A</th><th>unknown%</th></tr></tbody></table>"`;

exports[`comment tests buildComment when hiding coverage reports when not hiding unchanged when result diff is 0 then 0 diff in code coverage comment 1`] = `"<h2>Code Coverage</h2><table><tbody><tr><th>app A</th><th>50.00%</th><th>✅  0.00%</th></tr></tbody></table>"`;

exports[`comment tests buildComment when hiding coverage reports when not hiding unchanged when result diff is negative then negative diff in code coverage comment 1`] = `"<h2>Code Coverage</h2><table><tbody><tr><th>app A</th><th>50.00%</th><th>❌ ▾ -25.00%</th></tr></tbody></table>"`;

exports[`comment tests buildComment when hiding coverage reports when not hiding unchanged when result diff is positive then positive diff in code coverage comment 1`] = `"<h2>Code Coverage</h2><table><tbody><tr><th>app A</th><th>50.00%</th><th>✅ ▴ +25.00%</th></tr></tbody></table>"`;

exports[`comment tests buildComment when not hiding coverage reports when all results in one then return as expected 1`] = `
"<h2>Code Coverage</h2><table><tbody><tr><th>app diff 0</th><th>50.00%</th><th>✅  0.00%</th></tr></tbody></table> 

 <details><summary>Coverage Report</summary><table><tbody><tr><th>File                                </th><th> % Stmts </th><th> % Branch </th><th> % Funcs </th><th> % Lines </th><th> Uncovered Line #s </th></tr><tr><td>All&nbsp;files                           </td><td>&nbsp;  96.42 </td><td>&nbsp;      60 </td><td>&nbsp;  96.66 </td><td>&nbsp;  96.59 </td><td>&nbsp;                  </td></tr></tbody></table></details> <br/><table><tbody><tr><th>app diff -</th><th>50.00%</th><th>❌ ▾ -25.00%</th></tr></tbody></table> 

 <details><summary>Coverage Report</summary><table><tbody><tr><th>File                                </th><th> % Stmts </th><th> % Branch </th><th> % Funcs </th><th> % Lines </th><th> Uncovered Line #s </th></tr><tr><td>All&nbsp;files                           </td><td>&nbsp;  96.42 </td><td>&nbsp;      60 </td><td>&nbsp;  96.66 </td><td>&nbsp;  96.59 </td><td>&nbsp;                  </td></tr></tbody></table></details> <br/><table><tbody><tr><th>app diff +</th><th>50.00%</th><th>✅ ▴ +25.00%</th></tr></tbody></table> 

 <details><summary>Coverage Report</summary><table><tbody><tr><th>File                                </th><th> % Stmts </th><th> % Branch </th><th> % Funcs </th><th> % Lines </th><th> Uncovered Line #s </th></tr><tr><td>All&nbsp;files                           </td><td>&nbsp;  96.42 </td><td>&nbsp;      60 </td><td>&nbsp;  96.66 </td><td>&nbsp;  96.59 </td><td>&nbsp;                  </td></tr></tbody></table></details> <br/><table><tbody><tr><th>app diff null</th><th>unknown%</th></tr></tbody></table> 

 <details><summary>Coverage Report</summary><table><tbody><tr><th>File                                </th><th> % Stmts </th><th> % Branch </th><th> % Funcs </th><th> % Lines </th><th> Uncovered Line #s </th></tr><tr><td>All&nbsp;files                           </td><td>&nbsp;  96.42 </td><td>&nbsp;      60 </td><td>&nbsp;  96.66 </td><td>&nbsp;  96.59 </td><td>&nbsp;                  </td></tr></tbody></table></details> <br/>"
`;
